turbopuffer/namespace.py

```diff
@@ -199,178 +199,78 @@ def copy_from_namespace(self, source_namespace: str):
-            else:
-                raise ValueError('upsert() requires both ids= and vectors= be set.')
-        elif (ids is not None and attributes is None) or (attributes is not None and schema is None):
-            # Offset arguments to handle positional arguments case with no data field.
-            return self.upsert(VectorColumns(ids=data, vectors=ids, attributes=vectors), schema=attributes, distance_metric=distance_metric, encryption=encryption)
-        elif isinstance(data, VectorColumns):
-            payload = {**data.__dict__}
-
-            # Convert List[float] vectors to base64-encoded strings, if enabled.
-            if tpuf.upsert_vectors_as_base64 and payload["vectors"] is not None:
-                payload["vectors"] = [
-                    b64encode_vector(vector) if isinstance(vector, list) else vector
-                    for vector in payload["vectors"]
-                ]
-
-            if distance_metric is not None:
-                payload["distance_metric"] = distance_metric
-
-            if schema is not None:
-                payload["schema"] = schema
-            
-            if encryption is not None:
-                payload["encryption"] = encryption
-
-            response = self.backend.make_api_request('namespaces', self.name, payload=payload)
-
-            assert response.get('content', dict()).get('status', '') == 'OK', f'Invalid upsert() response: {response}'
-            self.metadata = None  # Invalidate cached metadata
-        elif isinstance(data, VectorRow):
-            raise ValueError('upsert() should be called on a list of vectors, got single vector.')
-        elif isinstance(data, list):
-            if isinstance(data[0], dict):
-                return self.upsert(VectorColumns.from_rows(data), schema=schema, distance_metric=distance_metric, encryption=encryption)
-            elif isinstance(data[0], VectorRow):
-                return self.upsert(VectorColumns.from_rows(data), schema=schema, distance_metric=distance_metric, encryption=encryption)
-            elif isinstance(data[0], VectorColumns):
-                for columns in data:
-                    self.upsert(columns, schema=schema, distance_metric=distance_metric, encryption=encryption)
-                return
-            else:
-                raise ValueError(f'Unsupported list data type: {type(data[0])}')
-        elif isinstance(data, dict):
-            if 'id' in data:
-                raise ValueError('upsert() should be called on a list of vectors, got single vector.')
-            elif 'ids' in data:
-                return self.upsert(VectorColumns.from_dict(data), schema=data.get('schema', None), distance_metric=distance_metric, encryption=encryption)
-            else:
-                raise ValueError('Provided dict is missing ids.')
-        elif 'pandas' in sys.modules and isinstance(data, sys.modules['pandas'].DataFrame):
```

Comment:
we lost the dataframe automatic integration. but it was purely sugar; no perf advantage compared to building the columns yourself

can always add it back if we see people asking